### PR TITLE
feature/CP-7024-ios-auto-request-notification-permission

### PR DESCRIPTION
**Pull request by cpandya25**

### CODE REVIEW

File: CleverPush/Source/CleverPushInstance.m [hunk 1]
```diff
@@ -385,8 +386,10 @@ - (void)initWithChannelId {
     if (subscriptionId != nil) {
         [self areNotificationsEnabled:^(BOOL notificationsEnabled) {
             if (!notificationsEnabled && !ignoreDisabledNotificationPermission) {
-                [CPLog info:@"notification authorization revoked, unsubscribing"];
-                [self unsubscribe];
+                if (autoRequestNotificationPermission) {
```
Comment (gbhrdt) [inline CleverPush/Source/CleverPushInstance.m:389]: Not required here for unsubscribe

File: CleverPush/Source/CleverPushInstance.m [hunk 2]
```diff
@@ -469,8 +472,10 @@ - (void)applicationWillEnterForeground API_AVAILABLE(ios(10.0)) {
                 [CPLog debug:@"CleverPushInstance: applicationWillEnterForeground: There is no subscription for CleverPush SDK."];
         } else {
             if (!notificationsEnabled && !ignoreDisabledNotificationPermission) {
-                [CPLog info:@"notification authorization revoked, unsubscribing"];
-                [self unsubscribe];
+                if (autoRequestNotificationPermission) {
```
Comment (gbhrdt) [inline CleverPush/Source/CleverPushInstance.m:475]: Not required here for unsubscribe

File: CleverPush/Source/CleverPushInstance.m [hunk 3]
```diff
@@ -885,54 +890,58 @@ - (void)subscribe:(CPHandleSubscribedBlock)subscribedBlock failure:(CPFailureBlo
         [center getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings *_Nonnull notificationSettings) {
             if (subscriptionId == nil && channelId != nil && notificationSettings.authorizationStatus == UNAuthorizationStatusNotDetermined) {
                 [self setConfirmAlertShown];
+            } else if (!autoRequestNotificationPermission && notificationSettings.authorizationStatus == UNAuthorizationStatusAuthorized ) {
+                [CleverPush setAutoRequestNotificationPermission:true];
```
Comment (gbhrdt) [inline CleverPush/Source/CleverPushInstance.m:894]: Why is this required?

File: CleverPush/Source/CleverPushInstance.m [hunk 4]
```diff
@@ -1116,7 +1125,7 @@ - (void)registerDeviceToken:(id)newDeviceToken onSuccess:(CPResultSuccessBlock)s
 
         if (@available(iOS 10.0, *)) {
             [[UNUserNotificationCenter currentNotificationCenter] getNotificationSettingsWithCompletionHandler:^(UNNotificationSettings* settings) {
-                if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) {
+                if (settings.authorizationStatus == UNAuthorizationStatusAuthorized || (!autoRequestNotificationPermission && settings.authorizationStatus == UNAuthorizationStatusNotDetermined)) {
```
Comment (gbhrdt) [inline CleverPush/Source/CleverPushInstance.m:1128]: Why is this condition required?

File: CleverPush/Source/CleverPushInstance.m [hunk 5]
```diff
@@ -1216,6 +1225,7 @@ - (void)setSyncSubscriptionRequestData:(NSMutableURLRequest*)request {
                                     isNil(country), @"country",
                                     isNil(timezone), @"timezone",
                                     isNil(language), @"language",
+                                    [NSNumber numberWithBool:autoRequestNotificationPermission], @"hasNotificationPermission",
```
Comment (gbhrdt) [inline CleverPush/Source/CleverPushInstance.m:1228]: Here we do not want to send the value of `autoRequest`. Instead we want to send the authorizationStatus - has user given the permission?

File: CleverPush/Source/CleverPushInstance.m [hunk 6]
```diff
@@ -1964,6 +1974,7 @@ - (void)startLiveActivity:(NSString*)activityId pushToken:(NSString*)token onSuc
                                  activityId, @"iosLiveActivityId",
                                  token, @"iosLiveActivityToken",
                                  subscriptionId, @"subscriptionId",
+                                 [NSNumber numberWithBool:autoRequestNotificationPermission], @"hasNotificationPermission",
```
Comment (gbhrdt) [inline CleverPush/Source/CleverPushInstance.m:1977]: Here we do not want to send the value of `autoRequest`. Instead we want to send the authorizationStatus - has user given the permission?

**Review by gbhrdt — CHANGES_REQUESTED** [review]

File: CleverPush/Source/CleverPushInstance.m (hunk 1 above)
Comment (cpandya25) [reply CleverPush/Source/CleverPushInstance.m:389]: Screenshot 2023-11-21 at 4 18 09 PM

Here is the scenario for that:

1. The user sets the `autoRequestNotificationPermission` to false.
2. We will call the sync subscription API, and we will get the subscription ID.
3. After that, the user will kill the application and launch it again.
4. Then it will go into this block because the user has not given the push notification permission, and we will get false from the `areNotificationsEnabled` method. That's why every second, if `autoRequestNotificationPermission` == false, the user will unsubscribe automatically.

That's why I kept the condition here. (Either every time a new subscription ID has been generated.)

File: CleverPush/Source/CleverPushInstance.m (hunk 2 above)
Comment (cpandya25) [reply CleverPush/Source/CleverPushInstance.m:475]: Same thing as above.

If the user sets the `autoRequestNotificationPermission` to false, they will minimize the application, and if the user relaunches the application, this function `applicationWillEnterForeground` will be called, and it will go into this block because the user has not given access to push notification permission, so it will return false for this `notificationsEnabled` method, and the `unsubscribe` function will be called.

Screenshot 2023-11-21 at 4 46 54 PM

File: CleverPush/Source/CleverPushInstance.m (hunk 3 above)
Comment (cpandya25) [reply CleverPush/Source/CleverPushInstance.m:894]: This is for if the user first sets the `autoRequestNotificationPermission` to false and then the user gives the notification permission access, then we need to set the flag `autoRequestNotificationPermission` to yes because if we don't set the true value, it will go into the false block every where we have to keep the condition like `if (!autoRequestNotificationPermission)`.

File: CleverPush/Source/CleverPushInstance.m (hunk 4 above)
Comment (cpandya25) [reply CleverPush/Source/CleverPushInstance.m:1128]: After getting the device token,

Earlier, it was like this:

`if (settings.authorizationStatus == UNAuthorizationStatusAuthorized) { }`
 if the notification permission had been granted, then only the subscribe API was being called.

But if the user had not given the notification permission and the `autoRequestNotificationPermission` flag was set to false, then we need to call the `subscribe()` method. That's why I keep the condition that if the user sets the `autoRequestNotificationPermission` to false and notification permission has not been asked for, then we need to call the `subscribe()` method.

Comment (gbhrdt) [reply CleverPush/Source/CleverPushInstance.m:1122]: No, in this case we do not want to call the subscribe method. The user can use setIgnoreDisabledNotificationPermission to enable this behaviour.
setAutoRequestNotificationPermission(false) should just NOT auto-request the permission - no other logic changes.

Comment (gbhrdt) [reply CleverPush/Source/CleverPushInstance.m:1122]: Please also go through the rest of the code and make sure no other logic changes when calling / not calling subscribe are involved in this PR.

**Review by gbhrdt — CHANGES_REQUESTED** [review]

**Review by gbhrdt — CHANGES_REQUESTED** [review]
